addons/bus/static/src/services/bus_monitoring_service.js

```diff
@@ -0,0 +1,110 @@
+ * lost connections. The bus connection is considered:
+ *
+ * - `LOST` if no connection could be established after a reconnect attempt.
+ * - `UNSTABLE` if the last connections was shorter than `STABLE_CONNECTION_TRESHOLD`.
+ * - `STABLE` once a connection reaches the `STABLE_CONNECTION_TRESHOLD`.
+ */
+export class BusMonitoringService {
+    connectionStableTimeout;
+    /** @type {CONNECTION_STATE[keyof CONNECTION_STATE]} */
+    connectionState;
+    isReconnecting = false;
+    lastConnectTs;
+    lastConnectionDuration;
+
+    constructor(env, services) {
+        const reactiveThis = reactive(this);
+        reactiveThis.setup(env, services);
+        return reactiveThis;
+    }
+
+    /**
+     * @param {import("@web/env").OdooEnv} env
+     * @param {Partial<import("services").Services>} services
+     */
+    setup(env, { bus_service }) {
+        bus_service.addEventListener("worker_state_updated", ({ detail }) =>
+            this.workerStateOnChange(detail)
+        );
+        browser.addEventListener("offline", () => {
+            // Reseting state when going offline to prevent false positive when
+            // waking up from sleep.
+            clearTimeout(this.connectionStableTimeout);
+            this.connectionState = null;
+            this.lastConnectionDuration = null;
+        });
+    }
+
+    /**
+     * Handle state changes for the WebSocket worker.
+     *
+     * @param {WORKER_STATE[keyof WORKER_STATE]} state
+     */
+    workerStateOnChange(state) {
+        if (!navigator.onLine) {
+            return;
+        }
+        switch (state) {
+            case WORKER_STATE.CONNECTING: {
+                this.isReconnecting = true;
+                break;
+            }
+            case WORKER_STATE.CONNECTED: {
+                this.isReconnecting = false;
+                this.lastConnectTs = DateTime.now().toMillis();
+                browser.clearTimeout(this.connectionStableTimeout);
+                this.connectionState =
+                    Number.isInteger(this.lastConnectionDuration) &&
+                    this.lastConnectionDuration < STABLE_CONNECTION_TRESHOLD
+                        ? CONNECTION_STATE.UNSTABLE
+                        : CONNECTION_STATE.STABLE;
+                if (this.connectionState === CONNECTION_STATE.UNSTABLE) {
+                    this.connectionStableTimeout = browser.setTimeout(
+                        () => (this.connectionState = CONNECTION_STATE.STABLE),
+                        STABLE_CONNECTION_TRESHOLD
+                    );
+                }
+                break;
+            }
+            case WORKER_STATE.DISCONNECTED: {
+                if (this.isReconnecting) {
+                    this.connectionState = CONNECTION_STATE.LOST;
+                    this.isReconnecting = false;
+                } else {
+                    this.lastConnectionDuration = DateTime.now().toMillis() - this.lastConnectTs;
+                }
+                break;
```

Comment:
I think `CONNECTION_STATE.UNSTABLE` is flawed: if we are `CONNECTED` for 2 days straight and we have some disconnects, the `UNSTABLE` will only show after 2 days (because `lastConnectionDuration` is last disconnect time - last connect, which was 2 days ago, this is necessarily greater than 2 minutes).